runtime/lua/vim/lsp.lua

```diff
@@ -558,7 +558,10 @@ function lsp.start_client(config)
     -- Buffer level cleanup
     vim.schedule(function()
       for _, bufnr in ipairs(active_buffers) do
-        lsp.diagnostic.clear(bufnr)
+        lsp.diagnostic.clear_diagnostic_cache(bufnr, client_id)
+        if vim.api.nvim_buf_is_valid(bufnr) then
+          lsp.diagnostic.clear(bufnr)
+        end
```

Comment:
I don't like that we have this copy and pasted twice effectively.

We can make `lsp.diagnostic.stop` or `lsp.diagnostic.cleanup` to clean this part up.

This could also make it so that we don't have to expose `clear_diagnostic_cache` since that's a private member that would be subject to change over time IMO

Reply:
I don't like it either. I'll make the change.